File: packages/lesspass-mobile/ios/LessPassModule.m
#import "LessPassModule.h"
#import <CommonCrypto/CommonHMAC.h>
#import <CommonCrypto/CommonCrypto.h>

@implementation LessPassModule

RCT_EXPORT_MODULE(LessPass);

RCT_EXPORT_METHOD(createFingerprint:(NSString *)key
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    NSString *data = @"";
    const char *cKey  = [key cStringUsingEncoding:NSUTF8StringEncoding];
    const char *cData = [data cStringUsingEncoding:NSUTF8StringEncoding];
    unsigned char cHMAC[CC_SHA256_DIGEST_LENGTH];
    
    CCHmac(kCCHmacAlgSHA256, cKey, strlen(cKey), cData, strlen(cData), cHMAC);
    
    NSMutableString *hash = [NSMutableString stringWithCapacity:CC_SHA256_DIGEST_LENGTH * 2];
    for (NSUInteger i = 0; i < CC_SHA256_DIGEST_LENGTH; i++) {
        [hash appendFormat:@"%02x", cHMAC[i]];
    }
    
    resolve(hash);
}

RCT_EXPORT_METHOD(calcEntropy:(NSString *)site
                    withLogin:(NSString *)login
           withMasterPassword:(NSString *)masterPassword
                 withCounter:(NSString *)counter
                     resolver:(RCTPromiseResolveBlock)resolve
                     rejecter:(RCTPromiseRejectBlock)reject)
{
    NSString *salt = [NSString stringWithFormat:@"%@%@%@", site, login, counter];
    NSData *keyData = [self pbkdf2WithPassword:masterPassword salt:salt keyByteCount:32 rounds:100000];
    
    if (keyData) {
        NSString *hexKey = [self hexStringFromData:keyData];
        resolve(hexKey);
    } else {
        reject(@"calcEntropy_error", @"Error computing entropy", nil);
    }
}

- (NSData *)pbkdf2WithPassword:(NSString *)password
                           salt:(NSString *)salt
                  keyByteCount:(NSUInteger)keyByteCount
                        rounds:(NSUInteger)rounds {
    NSData *saltData = [salt dataUsingEncoding:NSUTF8StringEncoding];
    NSMutableData *derivedKeyData = [NSMutableData dataWithLength:keyByteCount];
    
    int result = CCKeyDerivationPBKDF(kCCPBKDF2,
                                       [password UTF8String],
                                       [password lengthOfBytesUsingEncoding:NSUTF8StringEncoding],
                                       [saltData bytes],
                                       [saltData length],
                                       kCCPRFHmacAlgSHA256,
                                       (uint)rounds,
                                       [derivedKeyData mutableBytes],
                                       [derivedKeyData length]);
    
    if (result != kCCSuccess) {
        return nil;
    }
    
    return derivedKeyData;
}

- (NSString *)hexStringFromData:(NSData *)data {
    const unsigned char *dataBuffer = (const unsigned char *)[data bytes];
    if (!dataBuffer) {
        return [NSString string];
    }
    NSUInteger dataLength  = [data length];
    NSMutableString *hexString  = [NSMutableString stringWithCapacity:(dataLength * 2)];
    for (int i = 0; i < dataLength; ++i) {
        [hexString appendString:[NSString stringWithFormat:@"%02lx", (unsigned long)dataBuffer[i]]];
    }
    return [NSString stringWithString:hexString];
}

@end

File: mobile/ios/LessPassModule.swift
import Foundation

enum HMACAlgorithm {
    case MD5, SHA1, SHA224, SHA256, SHA384, SHA512

    func toCCHmacAlgorithm() -> CCHmacAlgorithm {
        var result: Int = 0
        switch self {
        case .MD5:
            result = kCCHmacAlgMD5
        case .SHA1:
            result = kCCHmacAlgSHA1
        case .SHA224:
            result = kCCHmacAlgSHA224
        case .SHA256:
            result = kCCHmacAlgSHA256
        case .SHA384:
            result = kCCHmacAlgSHA384
        case .SHA512:
            result = kCCHmacAlgSHA512
        }
        return CCHmacAlgorithm(result)
    }

    func digestLength() -> Int {
        var result: CInt = 0
        switch self {
        case .MD5:
            result = CC_MD5_DIGEST_LENGTH
        case .SHA1:
            result = CC_SHA1_DIGEST_LENGTH
        case .SHA224:
            result = CC_SHA224_DIGEST_LENGTH
        case .SHA256:
            result = CC_SHA256_DIGEST_LENGTH
        case .SHA384:
            result = CC_SHA384_DIGEST_LENGTH
        case .SHA512:
            result = CC_SHA512_DIGEST_LENGTH
        }
        return Int(result)
    }
}

extension String {
    func hmac(algorithm: HMACAlgorithm, key: String) -> String {
      let cKey = key.cString(using: String.Encoding.utf8)
      let cData = self.cString(using: String.Encoding.utf8)
      var result = [CUnsignedChar](repeating: 0, count: Int(algorithm.digestLength()))
      CCHmac(algorithm.toCCHmacAlgorithm(), cKey!, Int(strlen(cKey!)), cData!, Int(strlen(cData!)), &result)
      let hmacData:NSData = NSData(bytes: result, length: (Int(algorithm.digestLength())))
      var bytes = [UInt8](repeating:0, count: hmacData.length)
      hmacData.getBytes(&bytes, length: hmacData.length)
      var hexString = ""
      for byte in bytes {
          hexString += String(format:"%02hhx", UInt8(byte))
      }
      return hexString
    }
}

extension Data {
    struct HexEncodingOptions: OptionSet {
        let rawValue: Int
        static let upperCase = HexEncodingOptions(rawValue: 1 << 0)
    }

    func hexEncodedString(options: HexEncodingOptions = []) -> String {
        let hexDigits = Array((options.contains(.upperCase) ? "0123456789ABCDEF" : "0123456789abcdef").utf16)
        var chars: [unichar] = []
        chars.reserveCapacity(2 * count)
        for byte in self {
            chars.append(hexDigits[Int(byte / 16)])
            chars.append(hexDigits[Int(byte % 16)])
        }
        return String(utf16CodeUnits: chars, count: chars.count)
    }
}

func pbkdf2(hash: CCPBKDFAlgorithm, password: String, salt: String, keyByteCount: Int, rounds: Int) -> String? {
 
      guard let passwordData = password.data(using: .utf8), let saltData = salt.data(using: .utf8) else { return nil }
  
    var derivedKeyData = Data(repeating: 0, count: keyByteCount)
    let derivedCount = derivedKeyData.count
    let derivationStatus: Int32 = derivedKeyData.withUnsafeMutableBytes { derivedKeyBytes in
        let keyBuffer: UnsafeMutablePointer<UInt8> =
            derivedKeyBytes.baseAddress!.assumingMemoryBound(to: UInt8.self)
        return saltData.withUnsafeBytes { saltBytes -> Int32 in
            let saltBuffer: UnsafePointer<UInt8> = saltBytes.baseAddress!.assumingMemoryBound(to: UInt8.self)
            return CCKeyDerivationPBKDF(
                CCPBKDFAlgorithm(kCCPBKDF2),
                password,
                passwordData.count,
                saltBuffer,
                saltData.count,
                hash,
                UInt32(rounds),
                keyBuffer,
                derivedCount)
        }
    }
  return derivationStatus == kCCSuccess ? derivedKeyData.hexEncodedString() : nil
}

@objc(LessPassModule)
class LessPassModule: NSObject {
  @objc(createFingerprint:resolver:rejecter:)
  func createFingerprint(_ masterPassword: String,
                         resolver resolve: RCTPromiseResolveBlock,
                         rejecter reject:RCTPromiseRejectBlock) -> Void {
    resolve("".hmac(algorithm: HMACAlgorithm.SHA256, key: masterPassword))
  }
  
  @objc(calcEntropy:withLogin:withMasterPassword:withCounter:resolver:rejecter:)
  func calcEntropy(_ site: String,
                   withLogin login: String,
                   withMasterPassword masterPassword: String,
                   withCounter counter: String,
                   resolver resolve: RCTPromiseResolveBlock,
                   rejecter reject:RCTPromiseRejectBlock) -> Void {
    let salt = site + login + counter
    let r = pbkdf2(hash: CCPBKDFAlgorithm(kCCPRFHmacAlgSHA256), password: masterPassword, salt: salt, keyByteCount: 32, rounds: 100000)
    resolve(r)
  }
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
